File: C/Algorithm/Algorithm II/139. Word Break.c
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>

/*
 * Determines if a string can be segmented into words from a word dictionary.
 *
 * The 'wordBreak' function takes an input string 's', an array of words 'wordDict', and the size
 * of the word dictionary 'wordDictSize' as input. It checks if the input string 's' can be segmented
 * into words from the given word dictionary. It uses dynamic programming to track whether substrings
 * of 's' can be formed using words from the dictionary.
 *
 * Parameters:
 * - s: The input string to be segmented.
 * - wordDict: An array of words from the dictionary.
 * - wordDictSize: The number of words in the dictionary.
 *
 * Returns:
 * 'true' if 's' can be segmented into words from the dictionary, 'false' otherwise.
 */
bool wordBreak(char *s, char **wordDict, int wordDictSize)
{
  int STRING_LENGTH = strlen(s);
  int WORDS_LENGTH = wordDictSize;

  bool *isWordMatch = (bool *)malloc((STRING_LENGTH + 1) * sizeof(bool));
  memset(isWordMatch, false, (STRING_LENGTH + 1) * sizeof(bool));
  isWordMatch[0] = true;

  for (int i = 1; i <= STRING_LENGTH; i++)
  {
    for (int j = 0; j < WORDS_LENGTH; j++)
    {
      int WORD_LENGTH = strlen(wordDict[j]);

      if (i >= WORD_LENGTH && isWordMatch[i - WORD_LENGTH] && strncmp(s + i - WORD_LENGTH, wordDict[j], WORD_LENGTH) == 0)
      {
        isWordMatch[i] = true;
        break;
      }
    }
  }

  bool result = isWordMatch[STRING_LENGTH];
  free(isWordMatch);

  return result;
}

File: TypeScript/Algorithm/Algorithm II/139. Word Break.ts
function wordBreak(s: string, wordDict: string[]): boolean {
  const STRING_LENGTH: number = s.length;
  const WORDS_LENGTH: number = wordDict.length;

  const isWordMatch: boolean[] = new Array(STRING_LENGTH + 1).fill(false);
  isWordMatch[STRING_LENGTH] = true;

  for (let stringIndex: number = STRING_LENGTH - 1; stringIndex >= 0; stringIndex--) {
    for (let wordIndex: number = 0; wordIndex < WORDS_LENGTH; wordIndex++) {
      const WORD_LENGTH: number = wordDict[wordIndex].length;

      if (WORD_LENGTH + stringIndex <= STRING_LENGTH) {
        if (wordDict[wordIndex].localeCompare(s.substring(stringIndex, stringIndex + WORD_LENGTH)) === 0)
          isWordMatch[stringIndex] = isWordMatch[stringIndex + WORD_LENGTH];

        if (isWordMatch[stringIndex]) 
          break;
      }
    }
  }

  return isWordMatch[0];
}
